src/interpreter.cr

```diff
@@ -0,0 +1,95 @@
+
+struct Interpreter
+  @@shell : String | Nil
+
+  protected def self.shell : String
+    {% if flag?(:darwin) %}
+      @@shell ||= begin
+        Process.find_executable(ENV.fetch("RUBYSHELL", "dash")).as(String)
+      rescue TypeCastError
+        Process.find_executable(ENV.fetch("RUBYSHELL", "sh")).as(String)
+      end
+    {% else %}
+      @@shell ||= Process.find_executable(ENV.fetch("RUBYSHELL", "sh")).as(String)
+    {% end %}
+  end
+
+  # Ripped from Crystal::System::FileDescriptor#pipe, except we don't want `reader` to close-on-exec.
+  protected def self.pipe(command : String) : {IO::FileDescriptor, IO::FileDescriptor}
+    pipe_fds = uninitialized StaticArray(LibC::Int, 2)
+    raise IO::Error.from_errno("Could not create pipe") if LibC.pipe(pipe_fds) != 0
+
+    reader = IO::FileDescriptor.new(pipe_fds[0], false)
+    writer = IO::FileDescriptor.new(pipe_fds[1], false)
+    writer.close_on_exec = true
+    writer.sync = true
+    writer << command
+
+    {reader, writer}
+  end
+
+  # Ditto, except we only yield the path.
+  protected def self.pipe(command : String, &)
+    reader, writer = pipe(command)
+
+    begin
+      yield "/dev/fd/#{reader.fd}"
+    ensure
+      writer.flush
+      reader.close
+      writer.close
+    end
+  end
+
+  #######################################################################################
+
+  getter :name, :options, :script
+
+  @name : String = shell
+  @options : Array(String) = [] of String
+  @script : String = ""
+
+  def initialize(content : String)
+    match = /(\A#!.+\n|)([\S\s]*)\Z/.match(content).as(Regex::MatchData)
+    shebang = Process.parse_arguments(match[1])
+    shebang << self.class.shell unless shebang.any?
+
+    self.name = shebang
+    @script = match[2]
+  end
+
+  def exec(args : Array(String))
+    exit 0 if noop?
+
+    self.class.pipe(script) do |path|
+      Process.exec(name, [*options, path, *args])
+    end
+  end
+
+  def noop?
+    script.empty?
```

Comment:
Is this a case that might actually happen?

Reply:
The thought process here was that if you made a script with nothing but a shebang, the script is still valid, and running it always exits 0. That said, `noop?` (_whose name may need to change after this_) should require a shebang for it to be consider valid; whereas, a completely empty `command` is invalid.

Reply:
Actually, having revisited this, it's all by design as a result of my research and observation of [execve](https://man7.org/linux/man-pages/man2/execve.2.html). This is because the absence of a shebang on an executable file assumes `sh`. With that, plus what I'd mentioned above, that would be the expected behaviour in the event of the following examples:

```sh
> echo "" > empty
> echo "#!/usr/bin/env python3" > shebang_only
> chmod +x empty shebang_only
>
> ./empty; echo $?
0
> ./shebang_only; echo $?
0
```

```yaml
actions:
  empty:
    command: ""
    param_expansion: true
  shebang_only:
    command: "#!/usr/bin/env python3"
    param_expansion: true
```

```sh
> ops empty; echo $?
0
> ops shebang_only; echo $?
0
```
I'll add these examples to the `e2e` spec.